### PR TITLE
A fix for seeding the rng under older win GCC

### PR DESCRIPTION
**Pull request by bkentel**
- fix seeding as the title says.
- eliminate `djb2_hash`.
- fix a likely old bug where even if a seed was used, the rng would be seeded again anyway.

### CODE REVIEW

File: src/rng.cpp [hunk 1]
```diff
@@ -7,8 +8,12 @@ std::default_random_engine random_generator {std::random_device{}()};
 
 void seed_random_number_generator(char const* seed)
 {
-    random_generator.seed(
-        static_cast<unsigned>(djb2_hash(reinterpret_cast<unsigned char const*>(seed))));
+    if (!seed) {
+        // std::random_device returns the same sequence on older version of GCC on windows.
+        random_generator.seed(time(nullptr) ^ 0xDEADBEEF);
```
Comment (KA101) [inline src/rng.cpp]: time not declared in scope, sorry.  :-/

Comment (bkentel) [inline src/rng.cpp]: Sorry, should be ok now.

**Comment by KA101** [comment]
Can't test this whilst GCC refuses to compile, sorry that wasn't clear.

File: src/main.cpp [hunk 2]
```diff
@@ -39,6 +39,7 @@ int main(int argc, char *argv[])
 #endif
     bool verifyexit = false;
     bool check_all_mods = false;
+    bool explicit_seed = false;
```
Comment (kevingranade) [inline src/main.cpp]: The old flow where there was a seed string that was set and then passed to the seeding function was preferable.

File: src/rng.cpp (hunk 1 above)
Comment (kevingranade) [inline src/rng.cpp]: Don't get cute, just init with the value time() returns.

File: src/rng.cpp [hunk 3]
```diff
@@ -7,8 +9,12 @@ std::default_random_engine random_generator {std::random_device{}()};
...
+    } else {
+        random_generator.seed(std::hash<std::string> {}(seed));
```
Comment (kevingranade) [inline src/rng.cpp]: This is an unrelated change, it has no business being here.

File: src/text_snippets.cpp [hunk 4]
```diff
@@ -43,9 +43,9 @@ void snippet_library::add_snippet_from_json( const std::string &category, JsonOb
 
 int snippet_library::add_snippet(const std::string &category, const std::string &text)
 {
-    int hash = djb2_hash( (const unsigned char *)text.c_str() );
-    snippets.insert( std::pair<int, std::string>(hash, text) );
-    categories.insert( std::pair<std::string, int>(category, hash) );
+    auto const hash = static_cast<int>(std::hash<std::string> {}(text));
+    snippets.insert( make_pair(hash, text) );
+    categories.insert( make_pair(category, hash) );
```
Comment (kevingranade) [inline src/text_snippets.cpp]: I don't even know what you're thinking, the build is broken and your screwing around with random code that works fine already.

File: src/rng.cpp [hunk 5]
```diff
@@ -7,8 +9,12 @@ std::default_random_engine random_generator {std::random_device{}()};
 
 void seed_random_number_generator(char const* seed)
 {
-    random_generator.seed(
-        static_cast<unsigned>(djb2_hash(reinterpret_cast<unsigned char const*>(seed))));
+    if (!seed) {
+        // std::random_device returns the same sequence on older version of GCC on windows.
```
Comment (DavidKeaton) [inline src/rng.cpp]: I think this was in reference to my xor comment on seed initialization.

File: src/rng.cpp [hunk 6]
```diff
@@ -7,8 +9,12 @@ std::default_random_engine random_generator {std::random_device{}()};
...
+        random_generator.seed(time(nullptr) ^ 0xDEADBEEF);
```
Comment (DavidKeaton) [inline src/rng.cpp]: I think this was in reference to my xor comment on seed initialization.

File: src/rng.cpp (hunk 1 above)
Comment (bkentel) [inline src/rng.cpp]: It was @DavidKeaton -- I would probably have just used time() by itself otherwise. But no worries.

File: src/text_snippets.cpp (hunk 4 above)
Comment (BevapDin) [inline src/text_snippets.cpp]: @bkentel Just as a note: the hash of the snippet is stored in the saves along the item data, so a player always sees the same snippet when they look at the item description. Changing the hash function will have a negative effect on save  compatibility.

Comment (bkentel) [inline src/text_snippets.cpp]: @BevapDin Good information to have. A comment should probably be added to `djb2_hash` to indicate that its removal is basically only prevented by this.

Comment (kevingranade) [inline src/text_snippets.cpp]: A note in the snippets header perhaps, but there's no reason for the hash
function to reference things that use it, that's just asking to get out of
sync.

File: src/rng.cpp (hunk 6 above)
Comment (DavidKeaton) [inline src/rng.cpp]: It was worth a try, I liked it. :-)
